c8d: Add `--platform` flag to history, save and load (#5331)
docker/cli: docs/reference/commandline/image_save.md

```console
$ docker save -o ubuntu.tar ubuntu:lucid ubuntu:saucy
```

### <a name="platform"></a> Save a specific platform (--platform)

> The `--platform` option allows you to specify which platform-variant of the

Reviewer: ```suggestion
The `--platform` option allows you to specify which platform variant of the
```

image to save. By default, `docker save` saves all platform variants that
are present in the daemon's image store. Use the `--platform` option
to specify which platform variant of the image to save. An error is produced
if the given platform is not present in the local image store.

> The platform option takes the `os[/arch[/variant]]` format, for example,

Reviewer 2: nit:
```suggestion
The platform option takes the `os[/arch[/variant]]` format, for example
```

> `linux/amd64` or `limux/arm64/v8`. Architecture and variant are optional,

Reviewer 3: ```suggestion
`linux/amd64` or `linux/arm64/v8`. Architecture and variant are optional,
```

typo 3: return of the typo

and default to the daemon's native architecture if omitted.

The following examples pulls an `alpine` image for the daemon's native platform,

> then provides a custom platform, which fails until the given platform-variant

Reviewer 3: ```suggestion
then attempts to save it with a custom platform, which fails until the given platform-variant
```

> The following examples pulls an `alpine` image for the daemon's native platform,
> then provides a custom platform, which fails until the given platform-variant
> is pulled;

Reviewer 2: nit:
```suggestion
The following example: 
- pulls an `alpine` image for the daemon's native platform (because `--platform` is omitted);
- tries to save the image while providing a custom `--platform` value, which fails until the given platform-variant is pulled;
```

> The following examples pulls an `alpine` image for the daemon's native platform,
> then attempts to save it with a custom platform, which fails until the given
> platform-variant is pulled;
> 
> ```console
> $ docker pull alpine:latest
> Using default tag: latest
> latest: Pulling from library/alpine
> cf04c63912e1: Download complete
> Digest: sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d
> Status: Downloaded newer image for alpine:latest
> docker.io/library/alpine:latest
> 
> $ docker image save --platform=linux/s390x -o image.tar alpine:latest
> Error response from daemon: no suitable export target found for platform linux/s390x: no suitable image manifest found for platform {s390x linux  [] }
> 
> $ docker pull --platform=linux/s390x alpine
> Using default tag: latest
> latest: Pulling from library/alpine
> df110db6acd6: Download complete
> Digest: sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d
> Status: Image is up to date for alpine:latest
> docker.io/library/alpine:latest
> 
> $ docker image save --platform=linux/s390x -o image.tar alpine:latest
> ls -lh image.tar
> -rw-------  1 thajeztah  staff   3.9M Oct  7 11:06 image.tar
> ```

Reviewer: I prefer to lead with the happy case and to separate the happy and sad cases, like you did with `load`.

````suggestion
The following example pulls the RISC-V variant of the `alpine:latest` image and saves it to a tar archive.

```console
$ docker pull --platform=linux/riscv64 alpine:latest
latest: Pulling from library/alpine
8c4a05189a5f: Download complete 
Digest: sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d
Status: Downloaded newer image for alpine:latest
docker.io/library/alpine:latest

$ docker image save --platform=linux/riscv64 -o alpine_riscv.tar alpine:latest
```

The following example attempts to save a platform variant of `alpine:latest` that doesn't exist in the local image store, resulting in an error.

```console
$ docker image ls --filter reference=alpine --tree

alpine:latest               beefdbd8a1da       10.6MB         3.37MB       
├─ linux/riscv64            80cde017a105       10.6MB         3.37MB       
├─ linux/amd64              33735bd63cf8           0B             0B       
├─ linux/arm/v6             50f635c8b04d           0B             0B       
├─ linux/arm/v7             f2f82d424957           0B             0B       
├─ linux/arm64/v8           9cee2b382fe2           0B             0B       
├─ linux/386                b3e87f642f5c           0B             0B       
├─ linux/ppc64le            c7a6800e3dc5           0B             0B       
└─ linux/s390x              2b5b26e09ca2           0B             0B    

$ docker image save --platform=linux/amd64 -o alpine_amd64.tar alpine:latest
Error response from daemon: no suitable export target found for platform linux/s390x: no suitable image manifest found for platform {amd64 linux  [] }
```
````

Reviewer 4: Ah, yes, that works. I think I was originally working around some bugs we still had where things failed if the native platform wasn't present.

I tried to avoid `--tree` because the example _would_  work both with and without containerd image-store integration, but maybe we can add some wording to explain that output requires a multi-arch store, or perhaps skip it for now; WDYT?

Reviewer: If you don't want to show `--tree`, you could just stick with the `pull --platform=linux/riscv64` and save a different arch. But I think the tree output shows it nicely. Maybe a short note/tip would do...

```
> [!TIP]
> The `--tree` flag for `docker image ls` displays all locally available platform variants. (Requires the containerd image store.)
```

Reviewer: we should have a docs section for it... 🫥 😆

Reviewer 4: OK; I updated and included the tree output (YOLO); `--tree` still works without containerd enabled, but just doesn't show the architectures underneath (perhaps we should make it "work" and just show a single architecture 🤔)

Reviewer: SGTM
